### DISCUSSION TITLE
Seed database in production

### DISCUSSION THREAD

**Discussion by amrap030** [prisma/prisma#15890] [Q&A]
Hello together,

I am really unsure about the best strategy for seeding a database in production. Currently, I am seeding it right before the docker container starts, but this only works for the very first time, when the database is empty. After that, the seeding always fails.

So what would be the strategy here? Should database seeding be a manual step, or can we conditionally seed (if yes, how to conditionally check if seeding was already executed)?

Thank you in advance!

**Comment by Rackar** [comment]
I used a reinit script to do. Code here:

BE careful. All data will be droped. 

```js
async function dropAndInitDB() {
  await prisma.comment.deleteMany({})
  await prisma.post.deleteMany({})
  await prisma.category.deleteMany({})
  await prisma.fileBeenUploaded.deleteMany({})
  await prisma.theDocument.deleteMany({})
  await prisma.org.deleteMany({})
  await prisma.user.deleteMany({})


  let user = new UsersController(null)
  let tokenIns = await user.signup("admin", "admin", "123456")

  let userIns = await prisma.user.update({
    where: { username: "admin" },
    data: {
      roles: ["ADMIN"],
    },
  })

  let post = new PostController()
  let pIns = await post.create(
    "test1",
    "testetest1111",
    "etseastase aset a 1111",
    userIns.id
  )

  let comment = new CommentController()
  let cIns = await comment.create(pIns.data.id, "admin", "comment 1")
}
```

**Comment by nurul3101** [comment]
Hello @amrap030 👋 
I am trying to understand your use case of the need to do seeding in a production database. 
Ideally, seeding is helpful during the development phase, so it is okay to delete development data and seed it with other sample data. But for the production environment, the data could not be deleted. Is there any particular reason for seeding production database?

**Reply by amrap030** [reply]
@nurul3101 Yes on first startup of the application, I need some master data imported into the database from csv files. This data is not dynamic at the moment, i.e. there won‘t be anything added or deleted, it is only added once on startup. However, the data is necessary for the structure of the database to build relations for more efficient queries. If there is a different approach for such a use case, I would be glad to hear :)

**Reply by pantharshit00** [reply]
If you need absolute resiliency here(which might be overkill for this case), you can use a message queue. As soon as you will start your application, you will send a seed message to the message queue and it will sync all the data you need to your database and will be decoupled. 

Otherwise, you do need to invoke a function in your application when you will start it to seed the database or run a script before you start your application like I think you are doing that the moment.

**Reply by vladi-stevanovic** [reply]
:wave: Hello @amrap030 , I’m going through pending GitHub Discussions to check if you need any help. Is this issue still open?

Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! :raised_hands:

**Reply by amypellegrini** [reply]
Hi! I'm having exactly this problem. I have some data which is intended to be static. I can seed this data locally and everything works fine, however my app fails in a STAGE environment (didn't reach PROD obviously).

Is there guidance on how to do this?

Thanks!

**Reply by amypellegrini** [reply]
I mean, I could still run a seed command as part of the CI/CD pipeline, being very careful of not touching anything that doesn't need to be touched. I'm curious to know if there is a recommended approach or good practices around this.

**Reply by pantharshit00** [reply]
Hello @amypellegrini,

If seed data is important for functioning of your application, it should be part of your application startup process which should check for existence of this data. I will recommend incorporating this seed into the application code.

**Reply by revero-doug** [reply]
> Hello @amypellegrini,
> 
> If seed data is important for functioning of your application, it should be part of your application startup process which should check for existence of this data. I will recommend incorporating this seed into the application code.

semantically this still constitutes "seeding". most production apps would require some static data, so it's a bit confusing that seeding only applies to development environments in prisma land.

~~furthermore unless you pass a no-seed flag, seed is always run in development, so it's implied that your seed logic should be idempotent anyway.~~ nvm, seed only runs when running `prisma migrate reset` (or interactively resetting). `seed` shouldn't be relegated to test fixtures. I suspect there aren't more issues about this because people just choose to run it manually as the path of least resistance. this doesn't lend to great ergonomics if your application database requires any static data, irrespective of environment.

**Reply by nurul3101** [reply]
@revero-doug Would something like this be helpful for your use case?
- https://github.com/prisma/prisma/issues/12463

**Reply by revero-doug** [reply]
> @revero-doug Would something like this be helpful for your use case?
> 
> * [Environment dependent seeding #12463](https://github.com/prisma/prisma/issues/12463)

yes, just gave it a :+1: -- thanks!

**Reply by amypellegrini** [reply]
Thanks, looks good to me.

**Comment by justin-hackin** [comment]
I used a simple trick to determine when to seed: if the users table has no records, then the db needs seeding.

**Comment by SarenT** [comment]
I have a similar issue but it also applies for migrate. To generate the whole database structure, at some point, I have to run `npx prisma migrate deploy`. But the database service in my docker stack is not yet ready. So I have moved this process to the startup of the application service, which is not ideal.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
